VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+
+This VIP provides a specification for storing the available storage nodes on-chain.
+
+This is phase 2 of [VIP-7: Storage node discovery and selection (off-chain)](./vip-7.md)
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
+
+This method registers a new endpoint on the network:
+
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
+6. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
+
+Unregister a storage node from the network at the specified date:
+
+1. `didAddress`: The DID that is to be removed from the network
+2. `unregisterDatetime`: The unix timestamp of when the storage node should no longer be available for selection. Must be at least 28 days in the future.
+3. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.
```

Comment:
What's the motivation behind the 28-day notice (unregisterDatetime) if the `removeNode` is "just" disabling the discoverability?

Reply:
Good question. See latest commit where I've added more detail.

Reply:
But, as the note suggests the discoverability and the operation (being active and managing data) of the node are **not related**, the existing users of the node are not affected by the discoverability or not.

Having a notice period if the node is shutting down, yes, because users have to move their data.

But for chaining the discoverability, I don't see the point.
I'd argue, Storage Nodes may want to be non-discoverable instantly in some cases (technical issues, legal issues, ...) so they don't have more users coming in.

Reply:
I'm not following.

Can you suggest a change to illustrate what you're thinking?

Reply:
@tahpot, this passage:

> 2. `unregisterDatetime`: [...] will be removed from the network. [...] ensure users have sufficient time to migrate away [...]. [...] expect the node to become unavailable and their data to be deleted from `unregisterDatetime` onwards.

and this passage:

> Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.

are contradictory.

I personally lean towards the Note:

- Unregistering a node **only** disables its discoverability for new users. No notice period because it doesn't affects its users.
- Shutting down a node is a different process. It requires a notice period because users must move away.

Reply:
Hmm... I see what you mean, although I didn't intend it to come across as contradictory.

See the updated wording.

We don't want to support the concept of removing discoverability, but the node continues to operate for users. That will break the economic model when users are required to pay for storage via tokens. This contract is the source of truth of who is paying who.